composer/trainer/trainer.py

```diff
@@ -47,6 +47,35 @@
 
 log = logging.getLogger(__name__)
 
+def _dynamic_microbatch_wrapper(func: Callable) -> Callable:
+    """Wraps function to catch CUDA Out of Memory Errors and adaptively change microbatch
+    size if enabled to miaximize GPU usage.
+    """
+    def wrapper(*args, **kwargs):
+        self = args[0]
+        # Skip adaption if not enabled
+        if not self.adaptive_grad_accum:
+            return func
+        rerun = True
+        while rerun:
+            try:
+                rerun = False
+                return func(*args, **kwargs)
+            except RuntimeError as e:
+                if "CUDA out of memory" in str(e):
+                    # Raise runtime error if training 1 sample at a time still resulted in CUDA out of memory
+                    if self.state.grad_accum == self.state.batch_num_samples:
+                        raise RuntimeError(
+                            "CUDA out of memory. Train loop failed with an internal microbatch of size 1")
+                    else:
+                        rerun = True
+                        self.state.grad_accum = min(2 * self.state.grad_accum, self.state.batch_num_samples)
```

Comment:
Side thought, this would be a good place to check that `batch_num_samples` is still divisible by `grad_accum`. If not, can raise a `RuntimeError` telling the user this feature is only support for powers-of-two-sized batches.